src/smt/theory_datatype.cpp

```diff
@@ -410,6 +411,59 @@ namespace smt {
         return r;
     }
 
+    // explain the cycle root -> … -> app -> root
+    void theory_datatype::occurs_check_explain(enode * app, enode * const root) {
+        TRACE("datatype", tout << "occurs_check_explain " << mk_bounded_pp(app->get_owner(), get_manager()) << " <-> " << mk_bounded_pp(root->get_owner(), get_manager()) << "\n";);
+        enode* app_parent = nullptr;
+
+        while (app->get_root() != root->get_root()) {
+            theory_var v = app->get_root()->get_th_var(get_id());
+            SASSERT(v != null_theory_var);
+            v = m_find.find(v);
+            var_data * d = m_var_data[v];
+            SASSERT(d->m_constructor);
+            if (app != d->m_constructor)
+                m_used_eqs.push_back(enode_pair(app, d->m_constructor));
+            bool found = m_parent.find(app, app_parent);
```

Comment:
most maps have operator[] and unary "find" to access the map contents, asserting the key is in the domain.
This is different from stl semantics that inserts default elements.